perlock-core/src/main/java/name/mitterdorfer/perlock/impl/WatchServicePathWatcher.java

```diff
@@ -75,11 +83,13 @@ private void performRegistration() throws IOException {
      * Process all events for keys queued to the watcher
      */
     @Override
-    public void start() throws IOException {
+    public PathWatcher start() throws IOException {
         prepareWatcher();
         performRegistration();
         running = true;
+        lifecycleListener.onStart(this);
```

Comment:
> One thing I considered after submitting this PR was having the RunnablePathWatcherAdapter#start() method block until it actually starts.

I just thought about it. I think the easiest way to achieve this is by splitting the implementation in `WatchServicePathWatcher#start()`. `#start()` would just call `prepareWatcher()`and `performRegistration()` and set the flag `running` to `true`.

The adapter could then invoke the `start()` method synchronously:
```
public void start() {
            if (future != null) {
                throw new IllegalStateException("Cannot start a PathWatcher that is already running.");
            }
           // synchronous start
           delegate.start();
            LOG.trace("Submitting '{}' to executor service.", delegate);
            //submit itself when client wants to start watching. The pool will invoke the runnable when its ready
            future = executorService.submit(this);
        }
```

in the constructor of `RunnablePathWatcherAdapter` you have to store the concrete implementation class `WatchServicePathWatcher` as a field and in its `run()` method you just call `delegate.watch()` instead of `#start()`.